### DISCUSSION TITLE
Changing URL structure

### DISCUSSION THREAD

**Discussion by ghost** [FreshRSS/FreshRSS#5475] [Q&A] [answered]
Let's say I want to change the URL structure of the controller and action method from **example.com/i/?c=controller&a=action** to **example.com/i/controller/action**

I have understood that Minz framework handles all the URL requests.

So, i have tried understanding the code from Minz/Url.php and made the following changes.

Before :-

               if (isset($url['c'])
		 && $url['c'] != Minz_Request::defaultControllerName()) {
			$uri .= $separator . 'c=' . $url['c'];
			$separator = $and;
		}

		if (isset($url['a'])
		 && $url['a'] != Minz_Request::defaultActionName()) {
			$uri .= $separator . 'a=' . $url['a'];
			$separator = $and;
		}

		if (isset($url['params'])) {
			unset($url['params']['c']);
			unset($url['params']['a']);
			foreach ($url['params'] as $key => $param) {
				if (!is_string($key) || (!is_string($param) && !is_int($param))) {
					continue;
				}
				$uri .= $separator . urlencode($key) . '=' . urlencode((string)$param);
				$separator = $and;
			}
		}

After editing :- 

               if (isset($url['c'])
		 && $url['c'] != Minz_Request::defaultControllerName()) {
			$uri .= $url['c'] . '/';
		}

		if (isset($url['a'])
		 && $url['a'] != Minz_Request::defaultActionName()) {
			$uri .= $url['a'] . '/';
		}

		if (isset($url['params'])) {
			unset($url['params']['c']);
			unset($url['params']['a']);
			foreach ($url['params'] as $key => $param) {
				$uri .= urlencode($key). '/' . urlencode($param);
			}
		}

The code goes on printing the URL controller/action in the browser but always throws an error **The requested URL was not found on this server.**

Is there something I missed?

**Comment by Alkarex** [comment, marked as answer]
Hello,
I suggest you first make a working "hello world" example independently of FreshRSS, using only a plain PHP file.
Indeed, you first need to understand some more details about URLs, HTTP request, and Web server handlings (CGI).
The approach you have taken does not work, because this is not how a Web server is handling an HTTP request and interacting with PHP.

__First step__: try to make an example accepting an URL like `http://localhost/hello.php/one/two`

Here is an example of content to get you started with `hello.php`:

```php
<?php
phpinfo();
```

Hint: look at PHP's `$_SERVER['PATH_INFO']`

Note: this is something we use for our GReader API:

https://github.com/FreshRSS/FreshRSS/blob/723f7577d0a388a90779930754c5aacb9f66b168/p/api/greader.php#L988

__Second step (if relevant)__: try to update your example to accept an URL like `http://localhost/hello/one/two` (i.e. without the `.php` extension)

Hint: look at Apache's `ForceType`


__Alternative approach__: Try Apache's RewriteEngine to modify on the fly an URL like `http://localhost/hello/one/two` to `http://localhost/hello.php?a=one&b=two`

**Reply by ghost** [reply]
As per your suggestions, I have tried to go on with the RewriteEngine method and added this code to the .htaccess present in p folder.

```
<IfModule mod_rewrite.c>
   RewriteEngine On
   RewriteRule ^([0-9a-zA-Z]+)/$ ?c=$1 [NC,L]
</IfModule>
```
I expected the code to rewrite the url from **i/?c=subscription** to **i/subscription/** but produces no successful output.

**Reply by Alkarex** [reply]
I suggest that you first get your "hello world" example to work.

> I expected the code to rewrite the url from **i/?c=subscription** to **i/subscription/** but produces no successful output.

It is still not how this works. Your code only modifies an incoming URL to an internal representation.
You first have to try whether that code works (as it is, many links will break) when you manually type `i/subscription/` in the address bar, and if that works, then move on to change the links in the HTML sent to the client. There are a few ways to do so, including:
* Using another Apache feature to rewrite the links in the HTML responses, such as https://httpd.apache.org/docs/current/mod/mod_proxy_html.html#proxyhtmlurlmap
* Change the PHP code in charge of printing the URLs.
